// mongoose-id-assigner/lib/LocalStateStore.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("./utils");
class LocalStateStore {
    constructor() {
        this.collName = 'id_assigner';
        this.stateMap = new Map();
        LocalStateStore._instance = this;
    }
    getState(schema) {
        return this.stateMap.get(schema);
    }
    setState(schema, state) {
        const preState = this.stateMap.get(schema);
        if (!preState) {
            this.stateMap.set(schema, state);
            state.idAssigner.emit('unready');
        }
        else {
            this.stateMap.set(schema, state);
            if (state.readyState !== preState.readyState) {
                if (state.readyState === 2) {
                    state.idAssigner.emit('init');
                }
                if (state.readyState === 1) {
                    state.idAssigner.emit('ready');
                }
                if (state.readyState === 0) {
                    state.idAssigner.emit('unready');
                }
                if (state.readyState === -1) {
                    state.idAssigner.emit('error');
                }
            }
        }
    }
    getIdAssigner(schema) {
        const state = this.getState(schema);
        if (!state) {
            throw utils_1.PluginError(`schema's state not found in store.`);
        }
        return state.idAssigner;
    }
    clear() {
        this.stateMap.clear();
    }
    /**
     * Get the Collection name for all idAssigners
     * @return {string}
     */
    getCollName() {
        return this.collName;
    }
    /**
     * Can only be called on at app bootstrap, that is before any IA initialises
     * Well the functionality comes as setting a different collName will make ids
     * inconsistent for mongoose connection instances
     * @param {string} collName
     */
    setCollName(collName) {
        for (const config of this.stateMap.values()) {
            if (config.readyState === 1 || config.readyState === 2) {
                throw utils_1.PluginError('Could not Set CollName as Some Assigners have initialised, Call setCollName at app bootstrap level.');
            }
        }
        this.collName = collName;
    }
}
exports.LocalStateStore = LocalStateStore;
exports.localStateStore = new LocalStateStore();
