// mongoose-id-assigner/lib/MongooseIdAssigner.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const eventToPromise = require("event-to-promise");
const events_1 = require("events");
const mongodb_1 = require("mongodb");
const assigner_interfaces_1 = require("./assigner.interfaces");
const LocalStateStore_1 = require("./LocalStateStore");
const utils_1 = require("./utils");
const assign_fields_ids_1 = require("./utils/assign-fields-ids");
const configure_schema_1 = require("./utils/configure-schema");
const get_next_ids_1 = require("./utils/get-next-ids");
/**
 * The numbers also depict what you receive upon from readyState
 * @event 1. ready - If the init process is complete
 * @event 2. init - Initialising the idAssigner
 * @event 0. unready - Init process not started
 * @event -1. error - Init process error out
 */
class MongooseIdAssigner extends events_1.EventEmitter {
    constructor(schema, options) {
        super();
        if (!schema) {
            throw utils_1.PluginError('Schema for the IdAssigner Must be provided!');
        }
        if (!options || (options && !options.modelName)) {
            throw utils_1.PluginError('Plugin Options must be specified, with schema modelName!');
        }
        const modelName = options.modelName;
        if (LocalStateStore_1.localStateStore.getState(schema)) {
            throw utils_1.PluginError('Provided Schema already has an Assigner Instance!', modelName);
        }
        this.schema = schema;
        this.modelName = modelName;
        this.discriminatorKey = schema.get('discriminatorKey');
        this.retryTime = 20;
        this.retryMillis = 20; // after 20 millis
        this.options = utils_1.normaliseOptions(modelName, options);
        this.appendState({
            modelName: this.modelName,
            readyState: 0,
            idAssigner: this,
        });
        configure_schema_1.configureSchema(this);
    }
    get readyState() {
        return this.state.readyState;
    }
    get state() {
        return LocalStateStore_1.localStateStore.getState(this.schema);
    }
    get collection() {
        const { model, error } = this.state;
        if (!model) {
            if (error) {
                throw utils_1.PluginError('Cannot read Model, Error At Initialisation ' + error);
            }
            else {
                throw utils_1.PluginError('Cannot read Model, Not Initialised');
            }
        }
        return model.db.collection(LocalStateStore_1.localStateStore.getCollName());
    }
    static plugin(schema, options) {
        return new MongooseIdAssigner(schema, options);
    }
    refreshOptions() {
        return assign_fields_ids_1.refreshOptions(this);
    }
    getFieldConfig(field, discriminator) {
        if (discriminator && this.options.discriminators) {
            for (const [key, value] of this.options.discriminators.entries()) {
                if (key === discriminator) {
                    for (const [k, v] of value.entries()) {
                        if (k === field) {
                            return v;
                        }
                    }
                }
            }
        }
        if (this.options.fields) {
            for (const [k, v] of this.options.fields.entries()) {
                if (k === field) {
                    return v;
                }
            }
        }
        return;
    }
    getNextId(field, discriminator) {
        return __awaiter(this, void 0, void 0, function* () {
            const fieldConfig = this.getFieldConfig(field, discriminator);
            if (!fieldConfig) {
                throw utils_1.PluginError(`Requested Field, [${field}] does not have a Field Configuration!`, this.modelName);
            }
            switch (fieldConfig.type) {
                case assigner_interfaces_1.FieldConfigTypes.ObjectId:
                    return new mongodb_1.ObjectId();
                case assigner_interfaces_1.FieldConfigTypes.UUID:
                    return get_next_ids_1.getNextIdUUID(fieldConfig);
                case assigner_interfaces_1.FieldConfigTypes.Number:
                    yield this.refreshOptions();
                    return get_next_ids_1.getNextIdNumber(field, this, fieldConfig, '', 0, true);
                case assigner_interfaces_1.FieldConfigTypes.String:
                    yield this.refreshOptions();
                    return get_next_ids_1.getNextIdString(field, this, fieldConfig, '', 0, true);
            }
        });
    }
    appendState(state) {
        LocalStateStore_1.localStateStore.setState(this.schema, Object.assign(Object.assign({}, this.state), state));
    }
    initialise(modelInstance) {
        if (this.state.readyState === 0) {
            this.appendState({ readyState: 2 });
            return utils_1.initialiseOptions(modelInstance, this);
        }
        return this.state.readyState === 1
            ? Promise.resolve(1)
            : eventToPromise(this, 'ready').then(() => 1);
    }
}
exports.MongooseIdAssigner = MongooseIdAssigner;
