// mongoose-id-assigner/lib/utils/initialise-options.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const eventToPromise = require("event-to-promise");
const LocalStateStore_1 = require("../LocalStateStore");
const others_1 = require("./others");
const type_guards_1 = require("./type-guards");
function checkAndUpdateFieldConfigMap(fieldConfigMap, freshFieldConfigObj) {
    if (!freshFieldConfigObj || !fieldConfigMap) {
        return true;
    }
    let replace = true;
    for (const [field, config] of fieldConfigMap.entries()) {
        const oldConfig = freshFieldConfigObj[field];
        if (!oldConfig) {
            replace = true;
        }
        if (type_guards_1.isNumber(config) && type_guards_1.isNumber(oldConfig)) {
            if (oldConfig && config.nextId !== oldConfig.nextId) {
                replace = true;
                config.nextId = oldConfig.nextId;
            }
        }
        if (type_guards_1.isString(config) && type_guards_1.isString(oldConfig)) {
            if (oldConfig && config.nextId !== oldConfig.nextId) {
                replace = true;
                config.nextId = oldConfig.nextId;
            }
        }
        // fixme sort this issue out
        // throw error, or if dev tag, if reset true... then override
        // updated info on README.md until better fix, override
        if (config.type !== oldConfig.type) {
            replace = true;
        }
    }
    return replace;
}
// checks if fieldConfigs changed, update field nextIds only
// checks if AssignerOptions contains no field configs, replace = false
function checkAndUpdateOptions(options, freshOptions) {
    if (!freshOptions || (!freshOptions.fields && !freshOptions.discriminators)) {
        if (!options.fields && !options.discriminators) {
            return { abort: true, options };
        }
        else {
            options.timestamp = null;
            return { replace: true, options };
        }
    }
    // set timestamp
    options.timestamp = freshOptions.timestamp;
    // delete old options if new doesn't need it
    if (!options || (!options.fields && !options.discriminators)) {
        return { delete: true, options };
    }
    const rObject = { replace: true, options };
    rObject.replace = checkAndUpdateFieldConfigMap(options.fields, freshOptions.fields);
    // if discriminator options available
    if (options.discriminators) {
        for (const [dName, fieldConfigMap] of options.discriminators.entries()) {
            const replace = checkAndUpdateFieldConfigMap(fieldConfigMap, freshOptions ? freshOptions.discriminators[dName] : undefined);
            if (replace) {
                rObject.replace = true;
            }
        }
    }
    return rObject;
}
exports.checkAndUpdateOptions = checkAndUpdateOptions;
function refreshDBOptions(mongooseModel, idAssigner, retries = 0) {
    return __awaiter(this, void 0, void 0, function* () {
        const options = idAssigner.options;
        try {
            const freshOptions = yield mongooseModel.db
                .collection(LocalStateStore_1.localStateStore.getCollName())
                .findOne({ modelName: idAssigner.modelName });
            const mergedOptions = checkAndUpdateOptions(options, freshOptions);
            if (mergedOptions.abort) {
                idAssigner.appendState({
                    modelName: idAssigner.modelName,
                    readyState: 1,
                    model: mongooseModel,
                });
                return 1;
            }
            let update;
            if (mergedOptions.replace) {
                update = yield mongooseModel.db.collection(LocalStateStore_1.localStateStore.getCollName()).findOneAndReplace({
                    modelName: idAssigner.modelName,
                    timestamp: mergedOptions.options.timestamp,
                }, mergedOptions.options, {
                    upsert: true,
                });
            }
            else if (mergedOptions.delete) {
                update = yield mongooseModel.db.collection(LocalStateStore_1.localStateStore.getCollName()).findOneAndDelete({
                    modelName: idAssigner.modelName,
                    timestamp: mergedOptions.options.timestamp,
                });
                // new options requests deletion of old options
                // but those options have been updated by another process
                if (!update || !update.ok) {
                    return Promise.reject(others_1.PluginError('Error at initialisation, cannot delete old options, Still in use!', idAssigner.modelName));
                }
            }
            if (update && update.ok) {
                idAssigner.appendState({
                    readyState: 1,
                    model: mongooseModel,
                });
                return 1;
            }
            else {
                return Promise.reject(others_1.PluginError(`Initialisation error ${update}`, idAssigner.modelName));
            }
        }
        catch (e) {
            if (e.code === 11000) {
                if (retries > 30) {
                    throw others_1.PluginError('Initialisation error, maximum retries attained', idAssigner.modelName);
                }
                return refreshDBOptions(mongooseModel, idAssigner, ++retries);
            }
            return Promise.reject(e);
        }
    });
}
function dbInitialiseLogic(mongooseModel, idAssigner) {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            // create index, ensures no duplicates during upserts
            yield mongooseModel.db.collection(LocalStateStore_1.localStateStore.getCollName()).createIndex('modelName', {
                unique: true,
                background: false,
            });
            return yield refreshDBOptions(mongooseModel, idAssigner);
        }
        catch (e) {
            idAssigner.appendState({
                modelName: idAssigner.modelName,
                error: e,
                readyState: 3,
            });
            return 3;
        }
    });
}
function initialiseOptions(mongooseModel, idAssigner, retries = 0) {
    return __awaiter(this, void 0, void 0, function* () {
        const options = idAssigner.options;
        if (!options.network) {
            idAssigner.appendState({
                modelName: idAssigner.modelName,
                model: mongooseModel,
                readyState: 1,
            });
            return 1;
        }
        // connecting
        if (mongooseModel.db.readyState === 2) {
            return yield eventToPromise(mongooseModel.db, 'connected').then(() => dbInitialiseLogic(mongooseModel, idAssigner));
            // connected
        }
        else if (mongooseModel.db.readyState === 1) {
            return yield dbInitialiseLogic(mongooseModel, idAssigner);
        }
        if (retries < 10) {
            try {
                // 3 - disconnecting, wait more
                // 0 - disconnected, wait less as connection can be back anytime.
                yield others_1.waitPromise((mongooseModel.db.readyState === 3 ? 500 : 100) * retries);
                return initialiseOptions(mongooseModel, idAssigner, ++retries);
            }
            catch (e) {
                return Promise.reject(e);
            }
        }
        else {
            throw others_1.PluginError('Initialisation failed, cannot establish db connection not established!');
        }
    });
}
exports.initialiseOptions = initialiseOptions;
